### DISCUSSION TITLE
Google Summer of Code

### DISCUSSION THREAD

**Discussion by awvwgk** [fortran-lang/fpm#339] [Ideas]
Do we want to apply with @fortran-lang for Google Summer of Code projects with fpm?

We have plenty of open issues, some of them are already well discussed and fleshed out and could make a great self-contained project. Creating projects from our most important issues might help us to prioritize better on the development of fpm as well. Of course this is associated with a significant portion of work in preparation and over the summer.

See https://fortran-lang.discourse.group/t/fortran-monthly-call-january-2021/557 for the original idea.

**Comment by milancurcic** [comment]
Thank you for raising this. Yes, I think it's good time for us to apply this year for GSoC, and I agree, with a focus on fpm. I missed the call today and haven't watched the recording yet so I don't know to what extent has this been discussed already.

Applications for Orgs open Jan 29, and close Feb 19. We should have a clear idea about the project by Jan 29. Let's get on a call to flesh out ideas later this week or early next week.

@certik has experience with GSoC and I believe he will apply this year with LFortran.

I'll be happy to help with the Org application. Would anyone here like to be a mentor? Also, we have a few students who are actively involved with fortran-lang projects (I think @awvwgk, @lkedward, @ivan-pi). Does it make sense for them to apply as GSoC students (although they would make great mentors as well)? Or is it more appropriate to bring in a new student on board? 

Tagging also @everythingfunctional @urbanjost @arjenmarkus

**Reply by awvwgk** [reply]
Right, still student here. The GSoC guides only mentions that prior involvement in a project should be noted in the application. However, I wouldn't consider applying as student given my current involvement in this and other open source communities.

But, I am certainly in for mentoring a GSoC project with fpm.

**Reply by LKedward** [reply]
Unfortunately I'm no longer registered as a student since I took up a research position at the university while I ~procrastinate~ write-up my thesis. Happy to act as a mentor.

I haven't watched the call recording back yet, but +1 for a call to discuss this further.

**Comment by milancurcic** [comment]
It looks like we could have multiple projects, each with a student working on it. So it would make sense to apply with at least one project for fpm and another for stdlib.

**Reply by jvdp1** [reply]
> It looks like we could have multiple projects, each with a student working on it. So it would make sense to apply with at least one project for fpm and another for stdlib.

Indeed, it would make sense to propose one project for both fpm and stdlib. There is enough topics to work on in stdlib to find an attractive one ;)

**Comment by everythingfunctional** [comment]
I would absolutely be in favor of submitting fpm to GSoC. I wouldn't mind being a mentor either. Send me a calendar invite when you guys are planning to discuss.

**Comment by arjenmarkus** [comment]
Yesterday, during the monthly call I proposed to contact the organisers of
last year's FortranCon about recruiting students. I will do that this week.

Op wo 20 jan. 2021 om 04:28 schreef Brad Richardson <
notifications@github.com>:

**Comment by arjenmarkus** [comment]
I have contacted Alfio Lazzaro and Tiziano Müller about student involvement
from their side. Alfio suggested that I send the request to their CP2K
forum as well, as there we may find the actual students. I will do so later.

Op vr 22 jan. 2021 om 11:11 schreef Laurence Kedward <
notifications@github.com>:

**Comment by milancurcic** [comment]
I set up a Doodle poll for next week to discuss GSoC:

https://doodle.com/poll/4n9cwntktctkafdt?utm_source=poll&utm_medium=link

I plan to set a time Monday morning Eastern time at the latest.

**Comment by milancurcic** [comment]
Thank you all for responding on a short notice, we'll meet today at 2pm Eastern:

Milan Curcic is inviting you to a scheduled Zoom meeting.

Topic: Fortran-lang + GSoC
Time: Jan 25, 2021 02:00 PM Eastern Time (US and Canada)

Join Zoom Meeting
https://miami.zoom.us/j/97574635194?pwd=N2I0eHR3cnhJYTkzNUJSNGFGZFlwQT09

Meeting ID: 975 7463 5194
Passcode: 981002
One tap mobile
+16465588656,,97574635194#,,,,*981002# US (New York)

Dial by your location
        +1 646 558 8656 US (New York)
        +39 020 066 7245 Italy
        +39 021 241 28 823 Italy
        +39 069 480 6488 Italy
Meeting ID: 975 7463 5194
Passcode: 981002
Find your local number: https://miami.zoom.us/u/ad25uh4Rgp

Join by SIP
97574635194@zoomcrc.com

Join by H.323
162.255.37.11 (US West)
162.255.36.11 (US East)
213.19.144.110 (Amsterdam Netherlands)
213.244.140.110 (Germany)
103.122.166.55 (Australia)
64.211.144.160 (Brazil)
69.174.57.160 (Canada)
Meeting ID: 975 7463 5194
Passcode: 981002

**Comment by certik** [comment]
I won't be able to make the meeting, sorry about that. We should apply as fortran-lang. The most important part is a dedicated Ideas page with projects. New orgs will only get 2 slots.

**Comment by milancurcic** [comment]
Thank you @certik. We'll bounce _ideas_ ideas today and we should meet again next week to iterate on the application. I'll record the meeting for your reference.

Here are some reference docs from the guide that will be helpful for today:

* [Notes for 1st year orgs](https://google.github.io/gsocguides/mentor/notes-for-first-year-organizations)
* [Defining a project ideas list](https://google.github.io/gsocguides/mentor/defining-a-project-ideas-list)

**Comment by LKedward** [comment]
Apologies, it doesn't look like I'll be able to attend this meeting. I'll be sure to watch the recording and post any comments before the next meeting. Today I would have suggested #321 and #35 as a possible broad but self-contained project.

**Comment by milancurcic** [comment]
Here's a list of ideas that we discussed:

* fpm:
  - Local registry
  - Version constraint resolution
  - Preprocessor
  - Package manifest as a specification for testing
  - Source code distribution with foreign build systems + package distribution with conda-forge
* stdlib:
  - Strings
  - Containers
  - OS integration: POSIX and Windows API integration for file system and OS operations, subprocesses
* General
  - Standard conformance suite as an fpm package or a fpm feature (Sourcery Institute, @rouson)
  - Dockerfile / Docker image with fpm, stdlib, compilers and libraries set up and ready to use

Please post any other ideas that come up in this thread for now.

Tentative plan going forward:

1. Set up next week's meeting (I will do this)
2. In the meeting, discuss these ideas again and finalize a first draft of the list (but an idea can be added if it comes later)
3. Open an issue to track development and get feedback for each project idea. Find mentors for each project.

**Reply by awvwgk** [reply]
I updated the points about testing specifications and the source code distribution

**Comment by milancurcic** [comment]
Here's the recording of the meeting: https://youtu.be/13AcyCsHuQs

**Comment by certik** [comment]
Thanks everybody. I unfollowed the repository, so from now on I will only get emails when somebody mentions me, so I should be able to react sooner. For our next meeting, I would like @LKedward and myself to be able to attend.

The ideas above are a good start. @milancurcic can you start a wiki page? Then we can collectively update it with ideas.

**Reply by awvwgk** [reply]
I put the fpm related GSoC projects [here](https://github.com/fortran-lang/fpm/wiki/Google-Summer-of-Code).

**Reply by jvdp1** [reply]
Thank you @awvwgk . I'll do the same for `stdlib` [here](https://github.com/fortran-lang/stdlib/wiki/Google-Summer-of-Code) (Edit: done).

**Reply by certik** [reply]
We will apply as fortran-lang, so the ideas page must be just one page for fortran-lang with all related projects. I'll add lfortran projects there too. As many as we can. If we get accepted, then students will apply.

We need a dedicated channel for new students. Probably Discourse? If so, we also need clear instructions what students should do. 

The other thing to do is a "patch requirement". A student must send a patch to our repositories and the PR must be accepted, otherwise their application will not be considered. This ensures all candidates have enough skills to contribute.

This is how we do it for sympy:

https://github.com/sympy/sympy/wiki/GSoC-2019-Student-Instructions

https://github.com/sympy/sympy/wiki/GSoC-2019-Ideas

https://github.com/sympy/sympy/wiki/GSoC-2019-Application-Template

**Reply by milancurcic** [reply]
@certik considering that GitHub Wikis are per repo, we can use them for drafting and when done, consolidate all the projects on https://fortran-lang.org/gsoc-2021. It will look nicer than Wikis and should get more search results and hits.

> We need a dedicated channel for new students. Probably Discourse? If so, we also need clear instructions what students should do.

Do you mean for prospective students or once they started? I think Discourse or GitHub issues would both work.

> I'll add lfortran projects there too.

So LFortran won't apply through NumFocus? If not, would you like to be the admin for the fortran-lang application, considering you have the most experience with GSoC here?

**Reply by milancurcic** [reply]
I'm personally not in favor of the patch requirement, at least not in the first year. We already have against us a relatively small pool of applicants (I assume), that an additional filter may be unhelpful if not harmful. Applications area stressful enough for students already. I think the [You and Your Project section](https://github.com/sympy/sympy/wiki/GSoC-2019-Application-Template#you-and-your-project) already covers the ground for selecting good candidates.

**Reply by certik** [reply]
I will apply through as many organizations as I can. I will not apply to be a new organization just for LFortran, that chance is low. We have a decent chance as fortran-lang. We need more than 1 admin, yes I would like to be an admin, you should be too.

I disagree on the patch requirement, I am convinced we need it for projects that will contribute code to fpm or stdlib. If the student is unable or unwilling to send a patch, then the chance of a success of his or her project is extremely low and it will end up a waste of time of everybody, including us mentors. It is not worth.

**Comment by arjenmarkus** [comment]
Wrt the standard conformance, I suggested to use my collection of small
checking programs at
https://sourceforge.net/p/flibs/svncode/HEAD/tree/trunk/chkfeatures/.
The one thing to note (besides that the collection is not complete and some
features may be difficult to test explicitly) is that the programs are
designed to compile if the compiler supports the feature and to produce
some human-readable output to confirm the (correct) execution. They are not
meant perse to be used in a "probing" context, as you see in CMake of
autotools or the like.

Op ma 25 jan. 2021 om 21:09 schreef Milan Curcic <notifications@github.com>:

**Comment by milancurcic** [comment]
I set up a Doodle poll to meet next week. Ideally please mark your availability by end of day on Friday: https://doodle.com/poll/nkwic3cnpsc93zb2?utm_source=poll&utm_medium=link

Tagging @certik

**Reply by certik** [reply]
Done, thanks!

**Comment by arjenmarkus** [comment]
On the Intel Fortran forum, there was a question regarding automatic
documentation generation via FORD. Perhaps getting FORD into maintenance
again could be another project. I have no idea about the complexity and it
would involve parsing Fortran via Python, IIUIC, but it does seem useful.

Op do 28 jan. 2021 om 01:23 schreef Ondřej Čertík <notifications@github.com

**Reply by awvwgk** [reply]
Indeed, reviving FORD would be really awesome.

**Reply by certik** [reply]
Yes, we need a documentation generator. Unfortunately FORD is not permissively licensed. I very strongly suggest starting with a permissively licensed code.

**Reply by awvwgk** [reply]
I don't see any issue with the license on FORD. Starting from scratch to avoid contributing under a copyleft license seems harsh in my opinion.

**Reply by milancurcic** [reply]
Though it wouldn't be a show-stopper for GSoC (there are, of course, GPL'd projects that participate), I agree that it's in the long-term best interest of fortran-lang to have it permissively licensed.

Pinging @cmacmackin again (and others please upvote this message if you support this): **Would you please please consider multi-licensing FORD, so it has a permissive license beside GPL?** Many people would benefit from it being continually maintained and developed going forward.

**Reply by cmacmackin** [reply]
If you can explain to me why the GPL prevents people from developing it (I see no reason why it should unless people are just being bloody-minded) **and** find someone who is willing to formally take over as maintainer then I'll _consider_ relicensing.

**Reply by milancurcic** [reply]
People can develop it. I don't think that's the issue. 

The issue is that large orgs stay away from it, for reasons we may agree or disagree with. Because larger orgs stay away, smaller orgs stay away also. So it has less chance of growing through external contributions down the road.

So tiny orgs (like fortran-lang) have a choice: do we start with something workable that is less likely to grow down the road (short-term vision) or do we start from scratch and bet on it growing and being used more widely (long-term vision).

Ultimately, it's your project and your choice which I respect. But I will ask you every year. :)

**Reply by milancurcic** [reply]
> and find someone who is willing to formally take over as maintainer 

I agree this is important. I'm not the person to do it.

**Reply by cmacmackin** [reply]
Do you have any evidence that this has been the case with FORD? I find it hard to believe it's too much of a problem, given that the likes of Doxygen is gpl. As, indeed, is GCC.

**Reply by certik** [reply]
The above proposal, as I understood it, is to take over maintaining FORD and invest into developing it via GSoC via our fortran-lang community. So assuming that, then my suggestion is to invest into permissively licensed libraries, as we are talking about multi year long term effort. 

@cmacmackin please assume best intent on all participants here. When you write things like "...unless people are just being bloody-minded", it is not helpful. Nobody is bloody-minded here.

If you want one argument: I would like to be able to share code between FORD and other projects --- for example I plan to experiment with using LFortran parser for Fortran, and then handing it over to FORD to generate documentation. The GPL license prevents me to take any code from FORD and put it into LFortran's documentation backend.

**Reply by awvwgk** [reply]
Please keep in mind that open source licensing might be an emotionally loaded topic. Please try to keep this discussion civil.

> The issue is that large orgs stay away from it, for reasons we may agree or disagree with. Because larger orgs stay away, smaller orgs stay away also. So it has less chance of growing through external contributions down the road.

Many of our successful projects @grimme-lab are only available LGPL or GPL and it has not hindered the popularity nor the adoption even in big orgs so far. Also I see a healthy community of external contributors sharing their patches under a copyleft license with us. We have been urged in the past to relicense as well, but so far we always found an arrangement with the orgs to use the project under the existing licensing conditions.

**Reply by milancurcic** [reply]
I don't and it's quite possible that it didn't matter for FORD due to its small user base.

For Doxygen and GCC we can only speculate because we have no idea how they would evolve under a different license. Perhaps there are similar projects that changed license halfway and their development accelerated or not. But I don't know of such projects. 

But with fortran-lang I'd like to 100x FORD's use in 10 years (10x new Fortran projects times 10x the Fortran projects that use FORD), or whatever would be the new doc generator. But to do that we need every single factor on our side because we're not swimming upstream, but up a waterfall. Including the license. 

I don't think GPL is a show stopper. But I know that many people don't like it. 

License aside, would you like to mentor a GsOC student to develop FORD further? I can't think of a better person for it.

**Reply by cmacmackin** [reply]
If I understand your proposal properly, it could be achieved by writing an lfortran backend to Ford, which the license would allow. I don't see why you would need to include ford code in lfortran. If for some reason you do, one option would be to make it a separate release of lfortran and only cover that one with the gpl.

I fundamentally do not believe in permissive licenses. They won't stop me contributing to a project but I will not use them for my own codes. Frankly, I care more about free software than I do about any one project or, indeed, language. I therefore do not want my code used in anything proprietary. If that is truly vital to you then I'm afraid you'll have to write your own documentation generator. The parsing is by far the hardest part so it sounds like you're most of the way there.

My accusation of bloody-mindedness was directed at large companies that refuse to use gpl code despite their only intended use being permitted by the license. It also referred to people who might refuse to contribute meerge requests to gpl projects just because of the livrnse. It was not directed at those wishing to develop their own permissive libraries, whatever my disagreements with them might be.

**Reply by milancurcic** [reply]
> Many of our successful projects @grimme-lab are only available LGPL or GPL and it has not hindered the popularity nor the adoption even in big orgs so far.

That's great to hear. "Big org" is relative. Businesses that sell and distribute closed source software couldn't do this. Cloud computing and turning software into SaaS helps with GPL because software is no longer distributed.

My anecdote about licenses: My first GitHub project was GPL in 2013. There were a few requests here and there to multi-license. I liked the idea of GPL and these were small users so I said sorry. In 2015 two "big" users (team leads from Max Planck Institute and US Navy) asked to relicense. I thought, do I care about these users and my project being useful for them, or do I care about how GPL makes me feel? I relicensed.

What we see are people who ask to relicense. What we don't see are all the people that walk away without asking.

**Reply by milancurcic** [reply]
> I therefore do not want my code used in anything proprietary.

I think this hits the nail on the head and is the only argument that matters here, and I respect it.

**Reply by cmacmackin** [reply]
> License aside, would you like to mentor a GsOC student to develop FORD further? I can't think of a better person for it.

What would that entail from me? I don't know very much about GsOC. I might be able to do it through work.

**Reply by milancurcic** [reply]
I think about few hours per week to guide the student, during the summer. @certik knows more as he has experience with this. If you can, join the call on Tuesday and we'll discuss.

**Comment by milancurcic** [comment]
Thank you all for Doodle feedback. We'll meet on Tuesday (2/2) at 3 pm Eastern (20 UTC) to discuss GSoC projects. Zoom info below:

Milan Curcic is inviting you to a scheduled Zoom meeting.

Topic: Fortran-lang + GSoC
Time: Feb 2, 2021 03:00 PM Eastern Time (US and Canada)

Join Zoom Meeting
https://miami.zoom.us/j/98985693219?pwd=N040VlB6Rm9Ec3Z4QjE5eVg0bVVQZz09

Meeting ID: 989 8569 3219
Passcode: 182857
One tap mobile
+16465588656,,98985693219#,,,,*182857# US (New York)

Dial by your location
        +1 646 558 8656 US (New York)
        +39 020 066 7245 Italy
        +39 021 241 28 823 Italy
        +39 069 480 6488 Italy
Meeting ID: 989 8569 3219
Passcode: 182857
Find your local number: https://miami.zoom.us/u/adA1ClW89W

Join by SIP
98985693219@zoomcrc.com

Join by H.323
162.255.37.11 (US West)
162.255.36.11 (US East)
213.19.144.110 (Amsterdam Netherlands)
213.244.140.110 (Germany)
103.122.166.55 (Australia Sydney)
103.122.167.55 (Australia Melbourne)
64.211.144.160 (Brazil)
69.174.57.160 (Canada Toronto)
65.39.152.160 (Canada Vancouver)
Meeting ID: 989 8569 3219
Passcode: 182857

**Comment by arjenmarkus** [comment]
It seems this meeting will overlap another one (if I get my timezones
right), so I will probably not be able to join you.

Op ma 1 feb. 2021 om 04:28 schreef Milan Curcic <notifications@github.com>:

**Comment by urguru** [comment]
Hello.I am Guruprasad here
I am looking forward to apply for GSoC this year as a student and searching for various orgs to apply for.
I am totally new to this org. Can I join now and start my prep/contribution for GSoC as well as ooen source as a whole...?

**Reply by milancurcic** [reply]
Hi Guruprasad, yes, please join our call tomorrow if you'd like to meet some potential mentors for this Org, and to see what kind of projects you'd enjoy working on. Otherwise, follow this thread, we'll be posting updates as we flesh out project ideas.

**Comment by milancurcic** [comment]
@certik @awvwgk @LKedward @jvdp1 @everythingfunctional @rouson 

I copied the project ideas wikis here: https://github.com/fortran-lang/fortran-lang.org/wiki/Google-Summer-of-Code-2021

Any further edits please make there.

**Reply by awvwgk** [reply]
Thanks, I deleted the wiki page at the fpm repository to avoid potential confusion.

I'll try to collect ideas, notes and references to the fpm relates projects in the next days.

**Comment by milancurcic** [comment]
I started the discussion thread and GSoC tag on Discourse: https://fortran-lang.discourse.group/t/google-summer-of-code-2021/658
